### PR TITLE
ux-redesign: UserMessages redesign according to Patternfly

### PR DESCRIPTION
**Pull request by bond95**
Fixes: https://github.com/oVirt/ovirt-web-ui/issues/647

image
image
image

---
This change is [Reviewable](https://reviewable.io/reviews/ovirt/ovirt-web-ui/660)

### CODE REVIEW

**Comment by lizsurette** [comment]
@bond95 Sorry for my delayed review! Back from PTO :)

This is looking great! Just one comment:
1) Can you add an empty state when there are no notifications? I think it will help make it more clear. In this case the Clear All and Mark all Read actions can go away too. Check out an example here by clicking "Clear All": http://www.patternfly.org/pattern-library/communication/notification-drawer/#code

**Comment by bond95** [comment]
@lizsurette Done.
image
image

**Comment by bond95** [comment]
@gregsheremeta @sjd78 Please code review.

File: src/components/VmUserMessages/index.js [hunk 1]
```diff
@@ -3,120 +3,103 @@ import PropTypes from 'prop-types'
...
   }
 
   render () {
-    const { userMessages, onClearMessages } = this.props
+    const { userMessages, onClearMessages, onDisplayMessages, onDismissMessage } = this.props
 
     const idPrefix = `usermsgs`
-    const onToggle = () => {
-      this.setState({ show: !this.state.show })
+    const handleToggle = () => {
```
Comment (sjd78) [inline src/components/VmUserMessages/index.js]: If you move this to the class level (and rebind against this in the constructor), the arrow function will not need to be recreated at each render.

File: src/components/VmUserMessages/index.js [hunk 2]
```diff
@@ -3,120 +3,103 @@ import PropTypes from 'prop-types'
...
+      this.setState((prevState) => ({ show: !prevState.show }))
     }
 
-    let show = ''
-    if (this.state.show) {
-      show = 'show'
+    const handleExpand = () => {
```
Comment (sjd78) [inline src/components/VmUserMessages/index.js]: Same as `handleToggle`

File: src/components/VmUserMessages/index.js [hunk 3]
```diff
@@ -3,120 +3,103 @@ import PropTypes from 'prop-types'
...
-        <div className={`blank-slate-pf ${style['no-messages']}`}>
-          <div className='blank-slate-pf-icon'>
-            <span className='pficon pficon pficon-info' />
-          </div>
-          <h1>
-            {spaceToNonbreakableSpace(msg.noMessages())}
-          </h1>
-        </div>)
+    const filteredMessages = userMessages.get('records').filter((r) => r.get('show'))
+    const messagesList = filteredMessages.size ? filteredMessages.map(r => {
```
Comment (sjd78) [inline src/components/VmUserMessages/index.js]: Could be simplified:
````javascript
    let idCounter = 0
    const messagesList = filteredMessages.size
      ? filteredMessages.map(r =>
        <UserMessage
          key={`msg-${r.get('time')}`}
          record={r}
          id={`${idPrefix}-msg-${r.get('time')}-dropdown`}
          onDismissMessage={() => onDismissMessage(r.get('time'))}
        />
      )
      })
      : <NotificationDrawer.EmptyState title={msg.noMessages()} />
````

File: src/components/VmUserMessages/index.js [hunk 4]
```diff
@@ -3,120 +3,103 @@ import PropTypes from 'prop-types'
...
+        </NotificationDrawer>
       </li>
     )
   }
 }
 VmUserMessages.propTypes = {
   userMessages: PropTypes.object.isRequired,
   onClearMessages: PropTypes.func.isRequired,
+  onDisplayMessages: PropTypes.func.isRequired,
+  onDismissMessage: PropTypes.func.isRequired,
```
Comment (sjd78) [inline src/components/VmUserMessages/index.js]: What is the difference between "Clear" and "Dismiss"?

File: src/components/VmUserMessages/index.js [hunk 5]
```diff
@@ -3,120 +3,103 @@ import PropTypes from 'prop-types'
...
-        <pre className={style['message-box']} id={id}>
-          <Time time={record.time} cssClass={style['usermsg-time']} />
-          {record.message}
-        </pre>
-      </span>
-    </li>
-  )
+const UserMessage = ({ record, id, onDismissMessage }) => {
+  const time = getFormatedDateTime(record.get('time'))
+  return (<Notification seen>
```
Comment (sjd78) [inline src/components/VmUserMessages/index.js]: Where is `seen` defined?

**Comment by sjd78** [comment]
screenshot-localhost-3000-2018 08 05-15-52-54

In empty state, 3 questions.
1. Should the 'Clear all' button be disabled since it is a no-op anyway.
2. Should the 'Display all' button should also be disabled if there are no hidden messages? 
3. Why the extra space below the buttons?  I assume the buttons should be at the bottom of the min-height of the notification pane, or the pane should be shorter.

**Comment by lizsurette** [comment]
@sjd78 Adding my thoughts on some of the design-y questions :)

> If I "Clear All", should "Display All" bring them all back? (based on the code, yes)
I would even wonder if we need the concept of "Display All".

> If I hit "Clear All" now and then 20 minutes from now open the user messages and hit "Display All", should ALL of the previously generated messages come back? (based on the code, yes)
Again, what's the use case for needing to bring them back?

> Is "Dismiss" on a single message the same function as "Clear All"? (based on the code, yes)
It should just clear that one message.

**Comment by gregsheremeta** [comment]
> Again, what's the use case for needing to bring them back?

+1. Seems useless to me.

File: src/helpers.js [hunk 6]
```diff
@@ -217,3 +217,11 @@ export function arrayToMap (array, keySelector) {
     return accum
   }, {})
 }
+
+export function getFormatedDateTime (timestamp) {
```
[the remark this answers is not in the document]
Comment (gregsheremeta) [reply src/helpers.js]: +1. We can do this in a follow up PR

File: src/index-nomodules.css [hunk 7]
```diff
@@ -75,3 +75,13 @@ div.toolbar-pf {
 .popover {
   min-width: 120px;
 }
+
+.navbar-nav > .dropdown {
+  position: inherit;
+}
+
+.drawer-pf .blank-slate-pf .blank-slate-pf-title {
+    font-size: 1em;
```
[the remark this answers is not in the document]
Comment (gregsheremeta) [reply src/index-nomodules.css]: sometimes that's good practice for accessibility reasons

**Comment by gregsheremeta** [comment]
> Why the extra space below the buttons?

indeed, that should be shortened up

**Comment by sjd78** [comment]
@lizsurette , @gregsheremeta , @bond95  - the tray buttons match what is currently in webadmin (not that it makes sense either).

&lt;off-the-top-of-my-head-before-I-looked-at-the-patternfly-website-again&gt;
Dismiss (single), Clear all, and Display all could make more sense in terms of unread messages vs read messages (seen) and deleting the messages.  So the badge shows the unread count, the tray shows unread + read messages, and the bottom buttons could be "Hide all read" and "Show all read" or "Delete all read" and "Mark all unread" or something similar.  Read/unread/delete -- the operations that can be taken on messages.
&lt;/off-the-top-of-my-head&gt;

[PF Notification Drawer](https://www.patternfly.org/pattern-library/communication/notification-drawer/) has "Unread count", "Mark All Read", and "Clear All".  It also says "this option may be used differently across products", so there is just mild guidance.

File: src/index-nomodules.css (hunk 7 above)
[the remark this answers is not in the document]
Comment (sjd78) [reply src/index-nomodules.css]: I read all about font-size and relative units etc etc.  Percents and ems seem to work very similar as they both do an adjustment to the inherited height.  At this point, I have no real concerns, just curious.

File: src/helpers.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (sjd78) [reply src/helpers.js]: People will complain.  People don't like their dates in the "wrong" format.  In my i18n experience, it always comes up surprisingly quickly.

File: src/actions/userMessages.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import {
```
[the remark this answers is not in the document]
Comment (bond95) [reply src/actions/userMessages.js]: VmPortal have only one type of notifications, thats why notification drawer have only one category. Thus doing universal action like `addUserMessage` better in follow PR, for example here https://github.com/oVirt/ovirt-web-ui/issues/695

File: src/components/VmUserMessages/index.js (hunk 5 above)
Comment (bond95) [reply src/components/VmUserMessages/index.js]: There is not need to define `seen`, enough just add it as props.
P.S. Best practice for set bool props.

**Comment by bond95** [comment]
@sjd78 

> Why the extra space below the buttons? I assume the buttons should be at the bottom of the min-height of the notification pane, or the pane should be shorter.

Actually this is default behavior of notification drawer. Like here for example https://www.patternfly.org/pattern-library/communication/notification-drawer/#code
or in WebAdmin. So for consistent better to leave it as it is.

> I would even wonder if we need the concept of "Display All".
@sjd78 @lizsurette @gregsheremeta

**Comment by lizsurette** [comment]
> Actually this is default behavior of notification drawer. Like here for example https://www.patternfly.org/pattern-library/communication/notification-drawer/#code
> or in WebAdmin. So for consistent better to leave it as it is.

This could be a bug in PF-core. Taking a look at the PF-React example looks better to me without the extra spacing... https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html?selectedKind=patternfly-react%2FCommunication%2FNotification%20Drawer&selectedStory=Notification%20Drawer&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs



> @sjd78 @lizsurette @gregsheremeta Okay so you think we need to delete "Display All"?

Yes, I'm in favor of removing.

File: src/components/VmUserMessages/index.js (hunk 5 above)
Comment (sjd78) [reply src/components/VmUserMessages/index.js]: Ok, I got confused. You're just rendering all Notifications as seen.

At first glance, I assumed you intended to keep seen vs unseen (read vs unread).

**Comment by bond95** [comment]
I removed 'Display All' button. And made rebase.

image
image

**Comment by gregsheremeta** [comment]
> We could do better than Date.now() for a unique id for the messages

+1

> Get rid of `FAILED_EXTERNAL_ACTION` and `LOGIN_FAILED` and just replace them with a new `ADD_USER_MSG`

I don't know enough yet to get this one :)

> would make more sense from the data perspective to be `HIDE_ALL_USER_MSGS`, `HIDE_USER_MSG`, `SHOW_ALL_USER_MSGS`

I don't follow this one -- are we lining those up with something?

> Either they should be removed, or they should be updated on every operation.

+1

> @gregsheremeta - Should we add an issue for this so we don't forget?

Yep!

**Comment by gregsheremeta** [comment]
@sjd78  opened #712 #713

Can you elaborate on points 2 and 3 above?

**Pull request merged**